src/Compilers/Core/Portable/InternalUtilities/InterlockedOperations.cs

```diff
@@ -62,5 +63,29 @@ public static ImmutableArray<T> Initialize<T>(ref ImmutableArray<T> target, Immu
             var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref target, initializedValue, default(ImmutableArray<T>));
             return oldValue.IsDefault ? initializedValue : oldValue;
         }
+
+        /// <summary>
+        /// Initialize the immutable array referenced by <paramref name="target"/> in a thread-safe manner.
+        /// </summary>
+        /// <typeparam name="T">Elemental type of the array.</typeparam>
+        /// <param name="createArray">Callback to produce the array if <paramref name="target"/> is 'default'.  Will
+        /// only be called at most once.  Will not be called if 'target' is not 'default'.</param>
+        /// <returns>The value of <paramref name="target"/> after initialization.  If <paramref name="target"/> is
+        /// already initialized, that value value will be returned.</returns>
+        public static ImmutableArray<T> InterlockedInitialize<T>(ref ImmutableArray<T> target, Func<ImmutableArray<T>> createArray)
```

Comment:
Huh. I couldn't believe this wasn't an API that already existed. Feels like a BCL gap tbh.

Reply:
@sharwell for thoughts.

Reply:
This is exactly what I had in mind. It would be possible to call `Update` directly, but this helper is easier to read.